Python/import.c

```diff
@@ -303,10 +303,120 @@ _PyImport_Fini(void)
 PyObject *
 PyImport_GetModuleDict(void)
 {
-    PyInterpreterState *interp = PyThreadState_GET()->interp;
-    if (interp->modules == NULL)
-        Py_FatalError("PyImport_GetModuleDict: no module dictionary!");
-    return interp->modules;
+    PyObject *sysdict = PyThreadState_GET()->interp->sysdict;
+    if (sysdict == NULL)
+        Py_FatalError("PyImport_GetModuleDict: no sys module!");
+
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PyDict_GetItemId(sysdict, &PyId_modules);
+    if (modules == NULL)
+        Py_FatalError("lost sys.modules");
+    return modules;
+}
+
+/* In some corner cases it is important to be sure that the import
+   machinery has been initialized (or not cleaned up yet).  For
+   example, see issue #4236 and PyModule_Create2(). */
+
+int
+_PyImport_IsInitialized(PyInterpreterState *interp)
+{
+    if (interp->sysdict == NULL)
+        return 0;
+    _Py_IDENTIFIER(modules);
+    PyObject *modules = _PyDict_GetItemId(interp->sysdict, &PyId_modules);
+    if (modules == NULL)
+        return 0;
+    return 1;
+}
+
+PyObject *
+_PyImport_GetModule(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_Check(modules))
+        return PyDict_GetItem(modules, name);
+
+    PyObject *mod = PyObject_GetItem(modules, name);
+    // For backward-comaptibility we copy the behavior of PyDict_GetItem().
+    if (PyErr_Occurred())
+        PyErr_Clear();
+    Py_XDECREF(mod);
+    return mod;
+}
+
+PyObject *
+_PyImport_GetModuleWithError(PyObject *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
+    if (PyDict_Check(modules))
+        return PyDict_GetItemWithError(modules, name);
+
+    PyObject *mod = PyObject_GetItem(modules, name);
+    // For backward-comaptibility we copy the behavior
+    // of PyDict_GetItemWithError().
+    if (PyErr_Occurred() && !PyMapping_HasKey(modules, name))
+        PyErr_Clear();
+    return mod;
+}
+
+PyObject *
+_PyImport_GetModuleString(const char *name)
+{
+    PyObject *modules = PyImport_GetModuleDict();
```

Comment:
Hum, I would prefer to use PyUnicode_FromString() and call PyImport_GetModule() instead of duplicating the code, as you did in _PyImport_GetModuleId().

Reply:
I've dropped _PyImport_GetModuleString().